src/planscape/forsys/parse_forsys_output.py

```diff
@@ -353,3 +360,88 @@ def _append_ranked_project_to_scenario(
+    _geo_wkt_header: str
+
+    # The raw forsys output consists of 3 R dataframes.
+    # The "stand output" dataframe is converted into a dictionary of lists so
+    # that it's easier to process in Python.
+    _forsys_stand_output_df: dict[str, list]
+
+    def __init__(
+            self, raw_forsys_output: "rpy2.robjects.vectors.ListVector",
+            priority_weights: dict[str, float],
+            project_id_header: str, area_header: str, cost_header: str,
+            geo_wkt_header: str):
+        ForsysRankingOutputForASingleScenario.__init__(
+            self, raw_forsys_output, priority_weights, None, None,
+            project_id_header, area_header, cost_header)
+
+        self._save_raw_forsys_stand_output_as_dict(raw_forsys_output)
+        if geo_wkt_header not in self._forsys_stand_output_df.keys():
+            raise Exception(
+                "header, %s, is not a forsys output header" % geo_wkt_header)
+        self._geo_wkt_header = geo_wkt_header
+
+        project_area_geometries = self._get_project_area_geometries(
+            raw_forsys_output)
+        self._populate_geo_wkt_in_ranked_projects(project_area_geometries)
+
+    def _save_raw_forsys_stand_output_as_dict(
+            self, raw_forsys_output: "rpy2.robjects.vectors.DataFrame") -> None:
+        rdf = raw_forsys_output[self._STAND_OUTPUT_INDEX]
+        self._forsys_stand_output_df = {
+            key: np.asarray(rdf.rx2(key)) for key in rdf.names}
+
+    def _get_project_area_geometries(self,
+                                     raw_forsys_output: "rpy2.robjects.vectors.ListVector"
+                                     ) -> dict[int, Polygon | MultiPolygon]:
+        return self._merge_geos(
+            self._extract_geo_list(raw_forsys_output)
+        )
+
+    def _extract_geo_list(self,
+                          raw_forsys_output: "rpy2.robjects.vectors.ListVector"
+                          ) -> dict[int, list[Polygon]]:
+        rdf = raw_forsys_output[self._STAND_OUTPUT_INDEX]
+        stand_output_df = {
+            key: np.asarray(rdf.rx2(key)) for key in rdf.names}
+        print(stand_output_df.keys())
+        project_area_geometries = {}
+        for i in range(len(stand_output_df[self._project_id_header])):
+            id = stand_output_df[self._project_id_header][i]
+            geo = GEOSGeometry(stand_output_df[self._geo_wkt_header][i])
+            if id in project_area_geometries:
+                project_area_geometries[id].append(geo)
+            else:
+                project_area_geometries[id] = [geo]
+        return project_area_geometries
+
+    def _merge_geos(self,
+                    project_area_geometries: dict[int, list[Polygon]]
+                    ) -> dict[int, Polygon | MultiPolygon]:
+        merged_polygons = {}
+        for id in project_area_geometries.keys():
+            merged_polygons[id] = merge_polygons(
+                project_area_geometries[id], 0)
```

Comment:
note: I'm setting simplify_margin to 0 for now.

Other options I considered were None (no simplifying) and 300 * sqrt(2)/2 (remove pixel corners).

Reply:
sg!